modules/thesaurus/src/main/frontend/src/query/browse.jsx

```diff
@@ -0,0 +1,180 @@
+import BrowseTheme from "./browseStyle.jsx";
+
+import { REST_URL, MakeRequest, MakeChildrenFindingRequest } from "./util.jsx";
+
+class BrowseDialog extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      title: "Term browser",
+      lastKnownTerm: "",
+      parentNode: null,
+      currentNode: null,
+      childNodes: {},
+      childTermsToLookup: [],
+      hasChildren: {},
+      isExpanded: {},
+    };
+  }
+
+  // Construct a branch element for rendering
+  constructBranch = (id, name, ischildnode, defaultexpanded, bolded) => {
+    return(
+      <BrowseListChild
+        id={id}
+        name={name}
+        changeid={this.props.changeid}
+        registerinfo={this.props.registerinfo}
+        getinfo={this.props.getinfo}
+        expands={ischildnode}
+        defaultopen={defaultexpanded}
+        key={id}
+        headnode={!ischildnode}
+        bolded={bolded}
+      />
+    );
+  }
+
+  // Check for children of children elements
+  checkForChildren = (event, data, parent) => {
+    if (event === null) {
+      this.state.hasChildren[parent] = (data["rows"].length > 0);
+      this.forceUpdate();
+    } else {
+      console.log("Error: children lookup failed with code " + event.ToString());
+    }
+  }
+
+  // Callback from an onload to generate child nodes in state.childNodes
+  rebuildChildren = (event, data, parent) => {
+    if (event === null) {
+      // We have the children of our parent
+      this.state.childNodes[parent] = data["rows"].map((row, index) => {
+        // We also need to determine if this child has children of its own
+        MakeChildrenFindingRequest(row["id"], (status, data) => {this.checkForChildren(status, data, id)});
+        return this.constructBranch(row["id"], row["name"], true, false, false);
+      });
+    } else {
+      console.log("Error: term lookup failed with code " + event.ToString());
+    }
+  }
+
+  // Callback from an onload to generate the tree from a /suggest query about the parent
+  rebuildTree = (event, data) => {
+    if (event === null) {
+      // We have the node we're looking at, and its parent.
+      var currentNodeData = data["rows"][0];
+      var id = currentNodeData["id"];
+
+      // Look up every child of this node
+      MakeChildrenFindingRequest(id, (status, data) => {this.rebuildChildren(status, data, id)});
+
+      // Construct parent elements
+      const parentBranches = currentNodeData["parents"].map((row, index) => {
+        return this.constructBranch(row["id"], row["name"], false, false, false);
+      });
+
+      this.setState({
+        parentNode: parentBranches,
+        currentNode: this.constructBranch(id, currentNodeData["name"], true, true, true),
+        lastKnownTerm: id,
+      })
+    } else {
+      console.log("Error: initial term lookup failed with code " + event.ToString());
+    }
+  }
+
+  // Get suggestions and rebuild the browser tree for the given input
```

Comment:
We're not getting suggestions. We merely use the `/suggest` endpoint, but we're actually fetching all the details about the current node.

```suggestion
  // Rebuild the browser tree centered around the given term.
```